web-components/src/components/barcode-scanner/barcode-scanner.ts

```diff
@@ -0,0 +1,408 @@
+      this.codeReader = new BarcodeDetector()
+      this.detectFn = this.detectWithBarcodeDetector
+      this.sendBarcodeStateEvent({ state: BarcodeScannerState.DETECTOR_AVAILABLE })
+    } catch (error) {
+      this.sendBarcodeStateEvent({ state: BarcodeScannerState.DETECTOR_NOT_AVAILABLE })
+      console.warn("BarcodeDetector not available:", error)
+    }
+  }
+
+  private async detectWithBarcodeDetector(imageData: ImageBitmap): Promise<string | undefined> {
+    try {
+      // Type assertion for BarcodeDetector
+      const detector = this.codeReader
+      if (detector) {
+        const barcodes = await detector.detect(imageData)
+        if (barcodes.length > 0) {
+          return barcodes[0].rawValue
+        }
+      }
+    } catch (error) {
+      console.error("Error detecting barcode:", error)
+    }
+    return undefined
+  }
+
+  private async askPermission() {
+    try {
+      const permissionStatus = await navigator.permissions.query({ name: "camera" })
+
+      if (["granted", "prompt"].includes(permissionStatus.state)) {
+        await this.startCamera()
+      } else {
+        console.warn("Camera permission denied. Please enable it in your browser settings.")
+      }
+    } catch (error) {
+      console.error("Permission query error:", error)
+    }
+  }
+
+  private async startCamera() {
+    try {
+      this.stream = await navigator.mediaDevices.getUserMedia({
+        video: {
+          facingMode: "environment",
+          width: { ideal: 640, max: 1280 },
+          height: { ideal: 480, max: 720 },
+        },
+      })
+
+      if (this.video) {
+        this.video.srcObject = this.stream
+        await this.video.play()
+        this.sendBarcodeStateEvent({
+          state: BarcodeScannerState.STARTED,
+        })
+
+        // Set up video frame capture
+        this.setupFrameCapture()
+      }
+    } catch (err) {
+      console.error("Error accessing camera: ", err)
+    }
+  }
+
+  private stopVideo() {
+    if (this.stream) {
+      this.stream.getTracks().forEach((track) => track.stop())
+      // clean video
+      this.video.srcObject = null
+      this.stream = null
+    }
+  }
+
+  private sendBarcodeStateEvent(detail: { barcode?: string; state: BarcodeScannerState }) {
+    this.dispatchEvent(new CustomEvent(EventType.BARCODE_SCANNER_STATE, { detail }))
+  }
+
+  private exit(event: Event) {
+    event.stopPropagation()
+    this.stopVideo()
+    this.sendBarcodeStateEvent({
+      state: BarcodeScannerState.EXITED,
+    })
+  }
+
+  private setupFrameCapture() {
+    const video = this.video
+    const processFrame = async () => {
+      if (video.paused || video.ended) {
+        requestAnimationFrame(processFrame)
+        return
+      }
+
+      let imageBitmap: ImageBitmap | undefined
+      try {
+        imageBitmap = await createImageBitmap(video, 0, 0, video.videoWidth, video.videoHeight)
+      } catch (e) {
+        console.error("Error creating image bitmap:", e)
```

Comment:
As we run at each frame, we might want to avoid flooding the console, by only displaying this error maybe every 30 frames ?
(we could maintain a global variable with frameCount)